Fix Overlay of Combined Resources

janbartel opened the pull request:
Changes to the `Resource` class in jetty-12 lead to the breakage of war overlays for the jetty maven plugin, and also the scanning of directories when the base resource of a webapp is a `CombinedResource`.

janbartel committed:
Fix overlays with CombinedResource

janbartel committed:
Apply fix to ee9; remove comments in pom.xml

janbartel committed:
Add WEB-INF/classes to overlay

gregw commented on jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java:
```diff
@@ -581,31 +581,34 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
-                        parseClass(handlers, dirResource, classFile);
-                    }
-                    catch (Exception ex)
+        for (Resource r : dirResource)
+        {
+            Path dir = r.getPath();
+
+            try (Stream<Path> classStream = Files.walk(dir))
+            {
```
This fixes the issue, but does not really address the underlying problem.
We should not need to switch to Path API to walk a `Resource` tree.    `Resource` itself should have a `walk` method and then this code would not need to know about the plurality of the underlying resource.

gregw replied:
This fixes the issue, but not correctly.

By iterating over any combined resources, the parser might see multiple versions of the same class from different layers in an overlay, when it should only see the topmost one.
The solution is that we should not be using `Path.walk`.   Instead we need to implement a `Resource.walk` method that correctly handles duplicates in an overlay.

gregw reviewed (changes requested):
I know this was my suggested fix, but it is not correct.   We need to stay within the Resource abstraction.

gregw commented on jetty-ee10/jetty-ee10-maven-plugin/src/main/java/org/eclipse/jetty/ee10/maven/plugin/OverlayManager.java:
```diff
@@ -123,17 +127,18 @@ protected List<Overlay> getOverlays()
         //Get the name of the overlayed war and unpack it to a dir of the
         //same name in the temporary directory
-        String name = overlay.getResource().getFileName();
+        Path p = Paths.get(URIUtil.unwrapContainer(overlay.getResource().getURI()));
+        String name = p.getName(p.getNameCount() - 1).toString();
```
I don't think this is correct either, as if the overlay itself is a CombinedResource, then getURI will return null.    You are probably better just naming them "dir0", "dir1", ....
Or we can add a Resource.getSimpleName() method to extra a simple name that is not necessarily a file name.

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources

janbartel commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/CombinedResourceTest.java:
```diff
@@ -414,6 +415,52 @@ public void testContainsAndPathTo() throws Exception
 
+    @Test
+    public void testGetFileName()
+    {
```
I'd like a test that unambiguously shows what result you get for getFileName() when the resource is something like:  jar:file:/foo/bar.jar!/

gregw replied:
@janbartel see changed made to `MountPathResourceTest`

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -395,20 +398,16 @@ public static boolean isClassFile(Path path)
-        Path relative = base.relativize(path);
-
-        int count = relative.getNameCount();
+        int count = path.getNameCount();
```
This version of `FileID.isHidden()` isn't useful for anyone.

Only the `isHidden(Path base, Path path)` version is meaningful.

If we have a baseResource like `/path/.hidden/jetty/base` and a test path within it of `web/index.html`, then this should return false on `isHidden(base, path)`.

gregw replied:
All the two arg version was/is doing is getting the relative path and checking the segments.   This is just the version for code that already has the relative path.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -482,13 +495,13 @@ public static boolean isMetaInfVersions(Path path)
-        Path path0 = path.getName(0);
-        Path path1 = path.getName(1);
-        Path path2 = path.getName(2);
+        if (!"META-INF".equals(path.getName(0).toString()))
```
This implementation is probably fine, but I wonder, can this be lowercase?
Should this be a case-insensitive test?

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -482,13 +495,13 @@ public static boolean isMetaInfVersions(Path path)
-        return (path0.toString().equals("META-INF") &&
-            path1.toString().equals("versions") &&
-            path2.getFileName().toString().matches("[0-9]+"));
+        if (!"versions".equals(path.getName(1).toString()))
```
This implementation is probably fine, but I wonder, can this be uppercase?
Should this also be a case-insensitive test?

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -54,22 +54,66 @@ public static String dump(Resource resource)
+        if (otherPath == null)
+            return null;
+
+        return thisPath.relativize(otherPath);
```
If the Resources are on different FileSystems, this can throw an exception.

The safer way is to use the URIs and then relativize that way.

Eg: 

```
thisPath.toURI().relativize(otherPath.toURI()).toASCIIString();
```

But then what will you do with the URI path segment?
How can you create a new Resource from it?

Perhaps it's best to ensure that this `.getPathTo(Resource)` is only possible if the Resources also satisfy the `.contains(Resource)` logic?

joakime replied:
This also breaks when the two resources are in different locations.

Eg:

* Resource A is `/opt/www/jetty-base/`
* Resource B is `/home/user/tmp/index.html`

The result here is nonsense for our purposes.

Also doesn't work reliably if the the two Resources are different filesystem types.

Eg:

* Resource A is `jar:///path/to/lib.jar!/deep/`
* Resource B is `/home/user/tmp/index.html`

That will trigger an exception in the `thisPath.relativize(otherPath)`

You seem to have created this only for the AnnotationParser, seemingly because you didn't want to use the 2 argument FileId.isHidden() method?

gregw replied:
It is a useful method to have.    If resources are on different file systems, then a null return what we need.   I'll experiment with avoiding the exception and returning null in those situations.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:
```diff
@@ -252,24 +260,37 @@ public Iterator<Resource> iterator()
-        // Copy in reverse order
-        for (int r = _resources.size(); r-- > 0; )
+        Collection<Resource> all = getAllResources();
+        for (Resource r : all)
         {
-            _resources.get(r).copyTo(destination);
+            Path pathTo = getPathTo(r);
+            Path to = destination.resolve(pathTo);
+            if (r.isDirectory())
+            {
+                if (!Files.exists(to))
+                    Files.createDirectories(to);
+            }
+            else
+            {
+                Files.copy(r.getPath(), to, StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
+            }
```
This implementation is broken.

This shouldn't be doing the Path manipulation here.
The use of `getAllResources()` is wrong here, it should be delegating to each implementation's `Resource.copyTo(destination)` here instead.
That way cross-filesystem type copyTo is supported.

I have a testcase that breaks this implementation, and a fix in a branch locally I will be happy to demonstrate.

gregw replied:
@joakime The original code did do a delegation to each nested Resource in reverse order.   That is an algorithm that results in the correct result, but is very inefficient if large resources are overridden, since they will be copied and then replaced.
This algorithm avoid that double copy and the use of getAllResources is exactly the semantic we want to achieve.

But as you say,   we may need to duplicate the cross file system copy here.....      So either that code get's factored out to be usable by both impls of `copyTo` or we revert to the previous algorithm and live with the possible inefficiencies.

gregw replied:
I've fixed it with a cross FileSystem resolve technique

joakime reviewed (changes requested):
Out of 6 suspect APIs.
I've found 2 broken APIs so far.
Still working on writing tests for the other 4.

CombinedResource.copyTo(Resource)
  - test if destination is a different file system type (**BROKEN**, have fixed code)  

CombinedResource.contains(Resource other)
  - test if other is a different file system type

CombinedResource.getPathTo(Resource other)
  - test if other is a different file system type

Resource.contains(Resource other)
  - test if other is a different file system type
  - test if self is 1 JAR, and other is a different JAR

Resource.getPathTo(Resource other)
  - test if other is a different file system type (**BROKEN**, not possible to fix, revert?)

Resource.getAllResources()
  - test what happens if you have a symlink loop

joakime committed:
Fixes #10760 - Improved CombinedResource.copyTo and PathResource.copyTo

+ handles cross file system copy as well.

joakime committed:
Fixes #10760 - Improved Resource.contains and Resource.getPathTo

+ handles cross resource better.

joakime committed:
Fixes #10760 - Improved CombinedResource.contains

+ handles cross resource better.

joakime committed:
Fixes #10760 - Skip AnnotationParser candidate if not able to relativize path.

joakime committed:
Fixes #10760 - Add Resource.getAllResources testcase for symlink loops

joakime committed:
Fixes #10760 - Remove PathResource.TreeCopyFileVisitor (no longer used)

joakime commented:
I created a PR #10776 with extra testing (and fixes) for this PR.
@gregw need you to review that one too.

joakime committed:
Fixes #10760 - Checkstyle fix

joakime committed:
Fixes #10760 - Copy resources in reverse for proper Overlay behavior

joakime committed:
Fixes #10760 - Remove StringUtil.equals

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources
Reduced fallback to URI string manipulation.

gregw commented:
@joakime I merged your PR, but have subsequently modified it to reduce the usage of URI string manipulation, as that is very error prone (for example a `startWith` is not sufficient as you also need to check the next character is a '/').

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources
Reduced fallback to URI string manipulation.

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources
Reduced fallback to URI string manipulation.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -369,7 +372,7 @@ public static boolean isClassFile(Path path)
         String filename = fileNamePath.toString();
         // has to end in ".class"
-        if (!filename.toLowerCase(Locale.ENGLISH).endsWith(".class"))
+        if (!StringUtil.asciiEndsWithIgnoreCase(filename, ".class"))
```
Are `.CLASS` and `.Class` really valid class extensions?

gregw replied:
They are on DOS file systems.    I don't see we need to take the chance to make the distinction on other file systems

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -482,13 +495,13 @@ public static boolean isMetaInfVersions(Path path)
-        Path path0 = path.getName(0);
-        Path path1 = path.getName(1);
-        Path path2 = path.getName(2);
+        if (!StringUtil.asciiEqualsIgnoreCase("META-INF", path.getName(0).toString()))
```
Are `meta-inf` and `Meta-Inf` valid folder names?

gregw replied:
yes

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -482,13 +495,13 @@ public static boolean isMetaInfVersions(Path path)
-        return (path0.toString().equals("META-INF") &&
-            path1.toString().equals("versions") &&
-            path2.getFileName().toString().matches("[0-9]+"));
+        if (!StringUtil.asciiEqualsIgnoreCase("versions", path.getName(1).toString()))
```
Same as above: should the filename comparison be case-insensitive?

gregw replied:
See DOS file systems

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -516,13 +529,28 @@ public static boolean isNotMetaInfVersions(Path path)
-            return true;
+            return false;
+
+        return filenameSegment.toString().equalsIgnoreCase("module-info.class");
```
Same as above: should the filename comparison be case-insensitive?
And if it has to be, why not using `StringUtil.asciiEqualsIgnoreCase`?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:
```diff
@@ -252,24 +260,47 @@ public Iterator<Resource> iterator()
     {
-        List<Resource> result = new ArrayList<>();
-        for (Resource r : _resources)
+        Map<String, Resource> results = new TreeMap<>();
```
Why use a `TreeMap`? I don't think sorting by filename should be done here.

gregw replied:
Just to be deterministic rather than random

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java:
```diff
@@ -252,24 +260,47 @@ public Iterator<Resource> iterator()
+            }
+            else
+            {
+                Path pathFrom = r.getPath();
```
`r.getPath()` may return null. `Resource.newReadableByteChannel()` or `Resource.newInputStream()` should be used to perform the copy instead of the `Path` API.

gregw replied:
We have extra knowledge here that a non-directory resource will not be combined, so it should have a path.... but I guess that is future fragile. 
I'll make a utility copy that falls back to the inputStream version and use it here.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java:
```diff
@@ -204,6 +200,22 @@ public Path getPath()
+
+        Path thisPath = getPath();
+        if (thisPath == null)
+            throw new UnsupportedOperationException("Resources without a Path must implement contains");
```
Can a `PathResource` subclass return a null path?

gregw replied:
If it can, then it must implement contains

lorban commented on jetty-ee10/jetty-ee10-maven-plugin/src/main/java/org/eclipse/jetty/ee10/maven/plugin/OverlayManager.java:
```diff
@@ -123,17 +127,18 @@ protected List<Overlay> getOverlays()
         //Get the name of the overlayed war and unpack it to a dir of the
         //same name in the temporary directory
-        String name = overlay.getResource().getFileName();
+        Path p = Paths.get(URIUtil.unwrapContainer(overlay.getResource().getURI()));
```
Could you please modify the comment (or maybe add some form of assertion?) to say that the overlay's resource is expected to be of this kind: `jar:file:/some/file.war!/`?

lorban commented on jetty-ee9/jetty-ee9-annotations/src/main/java/org/eclipse/jetty/ee9/annotations/AnnotationParser.java:
```diff
@@ -578,31 +577,38 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
-                    }
-                });
+
+        for (Resource candidate : dirResource.getAllResources())
```
Is this testable?

gregw replied:
Yes, there are tests for getAllResources and Jan added an integration test that checked the bug that was in this loop that started this whole process off.

gregw committed:
Improvements to CombinedResource for AnnotationParser

Added `contains` and `getPathTo` methods to the Resource API, so they can be used by the AnnotationParser
Fixed numerous bugs in CombinedResource list and getAllResources
Reduced fallback to URI string manipulation.

joakime commented:
This branch is failing the following tests ...

```
[ERROR] Failures: 
[ERROR]   ResourceListingTest.testBasicResourceXHtmlListingDeep:101 expected: <true> but was: <false>
[ERROR]   ResourceListingTest.testBasicResourceXHtmlListingRoot:72 expected: <true> but was: <false>
[ERROR]   ResourceListingTest.testListingEncoding:293 expected: <true> but was: <false>
[ERROR]   ResourceListingTest.testListingFilenamesOnly:199 expected: <true> but was: <false>\
[ERROR]   ResourceListingTest.testListingProperUrlEncoding:238 expected: <true> but was: <false>
[ERROR]   ResourceListingTest.testListingWithQuestionMarks:269 expected: <true> but was: <false>
[ERROR]   ResourceListingTest.testResourceCollectionXHtmlListingContext:143 expected: <true> but was: <false>
```

Oddly, these are failing on `assertTrue(isValidXHtml(content));` tests.

joakime commented:
Ah, those are failing due to ...

```
java.io.IOException: Server returned HTTP response code: 429 for URL: http://www.w3.org/MarkUp/DTD/xhtml-object-1.mod
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:2000)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:677)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1397)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1333)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.startPE(XMLDTDScannerImpl.java:732)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.skipSeparator(XMLDTDScannerImpl.java:2101)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDecls(XMLDTDScannerImpl.java:2064)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDTDExternalSubset(XMLDTDScannerImpl.java:299)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1165)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1040)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:943)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:247)
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:342)
	at java.xml/javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:122)
	at org.eclipse.jetty.server@12.0.3-SNAPSHOT/org.eclipse.jetty.server.ResourceListingTest.isValidXHtml(ResourceListingTest.java:335)
	at org.eclipse.jetty.server@12.0.3-SNAPSHOT/org.eclipse.jetty.server.ResourceListingTest.testBasicResourceXHtmlListingDeep(ResourceListingTest.java:101)
```

Something @olamy is taking care of.
* https://github.com/jetty/jetty.toolchain/pull/298
* #10779

I'll merge this branch up from `jetty-12.0.x` HEAD now, it should correct these test failures.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-overlay-combined-resources

gregw commented:
Merging as CI failures are unrelated.
